src/client.jl

```diff
@@ -301,11 +301,13 @@ function processresponse!(client, conn, response, host, method, maintask, stream
         elseif stream && headerscomplete
             @log "processing the rest of response asynchronously"
             response.body.task = @async processresponse!(client, conn, response, host, method, maintask, false, tm, canonicalizeheaders, false)
-            return true, nothing
+            return true, StatusError(status(response), response)
```

Comment:
I don't think I understand the issue. Why were you ever catching a `nothing` exception? In this code path, we're returning `true` for successful, so there's no error?

Reply:
`status, err = (true, nothing)` is returned [from here](https://github.com/JuliaWeb/HTTP.jl/blob/master/src/client.jl#L304), [to here](https://github.com/JuliaWeb/HTTP.jl/blob/master/src/client.jl#L328).
But then, [down here](https://github.com/JuliaWeb/HTTP.jl/blob/master/src/client.jl#L342), depending on the status code, we do`throw(err)`, which is in this case `nothing`.

Note that [up here](https://github.com/JuliaWeb/HTTP.jl/blob/master/src/client.jl#L300) we seem to correctly return `(true, StatusError(...))`.

(Please excuse double post)

Reply:
It feels to me that the problem stems from code that implements a bunch of features in one place rather than separating them into distinct layers. (i.e. is seems wrong for `processresponse!()` to have knowledge of the possibility that the caller will need an exception object for a non-error result because the caller considers some status codes to be errors.

I'd like to see a seperate basic HTTP layer that has:
 - No default header values,
 - No interpretation of status codes,
 - No status code exceptions
 - No retries,
 - No redirects,
 - No cookies,
 - No authentication,

i.e there should be a layer that just sends a request and parses the response.

I think if that was done, then the higher order functions could be cleanly layered on top and the probability of pitfalls like this one would be greatly reduced. In fact each of the features listed above could be implemented as a seperate module, configurably, chained together, each with a `processresponse!()` function.

The user-level API could still enable a useful default set of high level features, nothing needs to change there.

See also [previous notes on layering](https://github.com/JuliaWeb/HTTP.jl/issues/63#issue-242916708)

Reply:
Great thoughts, but I'm just not sure it's really that practical. That's a great list for a "basic HTTP layer", but it's also a bit arbitrary. Why include error-handling at all? What about re-using open connections? Logging? Connect retries?

The point I'm getting at is the complete feature list that `client.jl` currently provides is even much longer than your list above, so where exactly do you draw the line in terms of "layers"? That seems like a lot of work for not a lot of gain to me. Right now, things are very configurable via the `RequestOptions` object; you can flip switches on and off as needed with negligible impact on performance and keeps all the code in one place, fairly easy to follow. I certainly looked into breaking things up into layers, but it just seemed like a bigger design challenge than is really worth once you get into the weeds.

Reply:
I would remove everything from the lowest layer that can be removed while leaving the ability to send a single request (header + body) and receive/parse a single response into header + body. This layer would assume that nothing goes wrong with the network. Dealing with network exceptions and deciding on retries vs rethrowing vs  throwing custom error types would be dealt with by higher layers.

Yes, absolutely retries and connection reuse I would implement as seperate layers. I would keep the layers thin and independently unit-testable. I would not combine multiple features into a layer unless they were unavoidably interdependent.

Logging I see as a service plane orthogonal to the layer stack, any layer should be able to log.

I've worked on a number of high performance network stacks that were rigorously factorised into layers. I don't think a layered approach implies a performance penalty. In fact the ability to configurably omit some layers completely can be a performance gain.

At the end of the day, any design that functions correctly and is easy test, easy to understand and easy to maintain is a good design. However, this particular bug smells to me like a sign that this design might benefit from more separation into layers and better defined interfaces between layers.